src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionResolver.cs

```diff
@@ -0,0 +1,239 @@
+                if (namespaces.Count == 0)
+                {
+                    break;
+                }
+
+                // Skip using directives 
+                var usingDirectiveNamespace = usingDirective.Statement.ParsedNamespace;
+                if (usingDirectiveNamespace.StartsWith("System") || usingDirectiveNamespace.Contains("="))
+                {
+                    continue;
+                }
+
+                // Insert all using directives that fit before the next using
+                if (namespaces.Peek().CompareTo(usingDirectiveNamespace) < 0)
+                {
+                    var usingDirectiveLineIndex = codeDocument.Source.Lines.GetLocation(usingDirective.Node.Span.Start).LineIndex;
+                    var head = new Position(usingDirectiveLineIndex, 0);
+                    var edit = new TextEdit() { Range = new Range(head, head), NewText = "" };
+                    do
+                    {
+                        edit.NewText += $"@using {namespaces.Dequeue()}{Environment.NewLine}";
+                    } while (namespaces.Count > 0 && namespaces.Peek().CompareTo(usingDirectiveNamespace) < 0);
+                    edits.Add(edit);
+                }
+            }
+
+            // Add the remaining usings to the end
+            if (namespaces.Count > 0)
+            {
+                var endIndex = usingDirectives.Last().Node.Span.End;
+                var lineIndex = GetLineIndexOrEnd(codeDocument, endIndex - 1) + 1;
+                var head = new Position(lineIndex, 0);
+                var edit = new TextEdit() { Range = new Range(head, head), NewText = "" };
+                do
+                {
+                    edit.NewText += $"@using {namespaces.Dequeue()}{Environment.NewLine}";
+                } while (namespaces.Count > 0);
+                edits.Add(edit);
+            }
+
+            return new WorkspaceEditDocumentChange(new TextDocumentEdit()
+            {
+                TextDocument = codeDocumentIdentifier,
+                Edits = edits,
+            });
+        }
+
+        private static WorkspaceEditDocumentChange GenerateUsingEditsAtTop(
+            RazorCodeDocument codeDocument,
+            VersionedTextDocumentIdentifier codeDocumentIdentifier,
+            string[] namespaceArray)
+        {
+            var namespaceList = namespaceArray.ToList();
+            namespaceList.Sort();
+
+            // If we don't have usings, insert after the last namespace or page directive, which ever comes later
+            var head = new Position(1, 0);
+            var lastNamespaceOrPageDirective = codeDocument.GetSyntaxTree().Root
+                .DescendantNodes()
+                .Where(n => IsNamespaceOrPageDirective(n))
+                .LastOrDefault();
+            if (lastNamespaceOrPageDirective != null)
+            {
+                var lineIndex = GetLineIndexOrEnd(codeDocument, lastNamespaceOrPageDirective.Span.End - 1) + 1;
```

Comment:
Why -1 then +1?

Reply:
This is weird but ended up being the cleanest/most general solution I could find. The directive span always terminates with then `\n` regardless of `\r\n` or `\n`, and so the previous character will always be on the last line of the directive's actual contents. Thus, the consequent line should be good for insertion. Sometimes this is outside the bound of the document if it ends with no newline, so we check that in `GetLineIndexOrEnd`.